### ISSUE TITLE
aria-labelledby rule does not check for invalid id values

### ISSUE THREAD

**Issue by Wildebrew** [dequelabs/axe-core#2620]
For id references, such as aria-labelledby or aria-describedby, the value being referenced must be a [valid id value](https://html.spec.whatwg.org/multipage/dom.html#the-id-attribute) (i.e. must not contain a white space).
I recently tested a page where the target of an aria-labelledby had an id attribute with a whitespace. The accessible name was not assigned (based on Chrome 84 with Jaws or NVDA), Wave reported an error but axe did not report a violation.

Expectation: axe should report a violation if the id reference is invalid (has white space or invalid characters)

Actual: axe does not report a violation.

Motivation: An invalid id value, when used as the target of labeling or informational reference constitues an accessibility issue (e.g. experienced by a screen reader user)

Example:
```<button aria-labelledby="my name"> </button> <span id="my name">Joe</span>```

axe-core version: Tested with latest axe Chrome plug-in

Browser and Assistive Technology versions
Jaws or NVDA with Chrome (not yet tested on other combinations but problem likely to persist).

**Comment by straker** [comment]
The [aria-labelledby attribute](https://www.w3.org/TR/wai-aria-1.1/#aria-labelledby) can [use multiple labels](https://www.w3.org/TR/wai-aria-1.1/#valuetype_idref_list), so a space separated list of IDs is a valid value. As for your test example, axe-core does throw a violation of `aria-valid-attr-value` on the `<button>` because the IDs do not exist in the DOM (the `<span>` does not have a valid ID).

**Comment by Wildebrew** [comment]
Here's the exact location I noticed this:

1. go to https://www.truistinsurance.com/
2. Expand the "about us" menu in the header
3. Run axe

My axe shows up with "Congratulations, no issues", but the navigation
region that contains the "about us" list of links does not have an
accessible name (it is not announced by a screen reader).

If you inspect the nav element around the expanded menu, it has
aria-labelledby="About Us", the button you click to expand it has
id="About Us".

So, they match but the button has a space in it.
I didn't see any elements with id="About" or id="Us" so there should
be either one or two broken references in the aria-labelledby value.

Maybe something else is at play here, I have to dive deeper. And,
you're right, I didn't think about the multiple id reference thing for
aria-labelledby.

Still, in that scenario it should return two accounts of broken
references (no matching "About" and "Us" id values).

This is ultimately more of a Linter type of situation, but invalid ID
attributes can play a star role in broken accessibility.



On 11/3/20, Steven Lambert <notifications@github.com> wrote:
> The [aria-labelledby
> attribute](https://www.w3.org/TR/wai-aria-1.1/#aria-labelledby) can [use
> multiple labels](https://www.w3.org/TR/wai-aria-1.1/#valuetype_idref_list),
> so a space separated list of IDs is a valid value. As for your test example,
> axe-core does throw a violation of `aria-valid-attr-value` on the `<button>`
> because the IDs do not exist in the DOM (the `<span>` does not have a valid
> ID).
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dequelabs/axe-core/issues/2620#issuecomment-721194761


-- 
Work hard. Have fun. Make history.

**Comment by straker** [comment]
Interesting. I get the violation in the axe extension as well

![image](https://user-images.githubusercontent.com/2433219/98008891-370e4b80-1db2-11eb-8a10-82d9348d3cac.png)

**Comment by straker** [comment]
I do notice that we don't specifically call out that the reason the value is invalid is because the IDs are not present on the page. We could probably improve our message in this case as the values are indeed value on their own.

**Comment by Wildebrew** [comment]
I am using axe for Chrome v4.6.1 (based on axe Core 4.0.2).
Is this the correct version (I am still getting 0 violations detected).
If it is, there's something off with my installation (and scine this
repository is not for addressing installation issues I'll try to
figure that out offline).

On 11/3/20, Steven Lambert <notifications@github.com> wrote:
> I do notice that we don't specifically call out that the reason the value is
> invalid is because the IDs are not present on the page. We could probably
> improve our message in this case as the values are indeed value on their
> own.
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dequelabs/axe-core/issues/2620#issuecomment-721217123


-- 
Work hard. Have fun. Make history.

**Comment by WilcoFiers** [comment]
@Wildebrew We're not able to reproduce the scenario. However we run it, axe-core reports multiple issues, not just the aria-labelledby one either. Any thought how you can come to 0 issues on this? Are you sure you're running this in the correct state?

**Comment by straker** [comment]
Closing as we were not able to reproduce the problem.
